File: smalivm/src/main/java/org/cf/smalivm/ObjectInstantiator.java
package org.cf.smalivm;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.common.base.Defaults;

public class ObjectInstantiator {

    private static final Logger log = LoggerFactory.getLogger(ObjectInstantiator.class.getSimpleName());

    public static Object newInstance(Class<?> klazz) throws InstantiationException, IllegalAccessException,
                    IllegalArgumentException, InvocationTargetException {
        if (Enum.class.isAssignableFrom(klazz)) {
            return newEnumInstance(klazz);
        }

        Object newInstance;
        try {
            // Try it without any funny business.
            newInstance = klazz.newInstance();
        } catch (InstantiationException | IllegalAccessException e) {
            if (log.isTraceEnabled()) {
                log.trace("{} has no default constructor, picking another", klazz);
            }

            // Alright, we'll try it the hard way.
            Constructor<?>[] ctors = klazz.getDeclaredConstructors();
            Constructor<?> ctor = ctors[0];
            ctor.setAccessible(true); // the little games we play, java...
            Object[] args = new Object[ctor.getParameterCount()];
            Class<?>[] parameterTypes = ctor.getParameterTypes();
            for (int i = 0; i < args.length; i++) {
                Class<?> parameterType = parameterTypes[i];
                args[i] = Defaults.defaultValue(parameterType);
            }
            newInstance = ctor.newInstance(args);
        }

        return newInstance;
    }

    @SuppressWarnings({ "unchecked", "rawtypes" })
    private static Object newEnumInstance(Class<?> klazz) {
        return Enum.valueOf((Class<? extends Enum>) klazz, "$shadow_instance");
    }

}

File: smalivm/src/main/java/org/cf/smalivm/ObjectInstantiator.kt
package org.cf.smalivm

import com.google.common.base.Defaults
import org.slf4j.LoggerFactory
import java.lang.reflect.InvocationTargetException

object ObjectInstantiator {
    private val log = LoggerFactory.getLogger(ObjectInstantiator::class.java.simpleName)

    @Throws(InstantiationException::class, IllegalAccessException::class, IllegalArgumentException::class, InvocationTargetException::class)
    fun newInstance(klazz: Class<*>): Any {
        if (Enum::class.java.isAssignableFrom(klazz)) {
            return newEnumInstance(klazz)
        }
        var newInstance: Any
        try {
            // Try it without any funny business.
            newInstance = klazz.newInstance()
        } catch (e: Exception) {
            if (e !is InstantiationException || e !is IllegalAccessException) {
                throw e
            }
            log.trace("{} has no default constructor, picking another", klazz)

            // Alright, we'll try it the hard way.
            val ctor = klazz.declaredConstructors[0]
            ctor.isAccessible = true // the little games we play, java...
            val args = arrayOfNulls<Any>(ctor.parameterCount)
            val parameterTypes = ctor.parameterTypes
            var i = 0
            while (i < args.size) {
                val parameterType = parameterTypes[i]
                args[i] = Defaults.defaultValue(parameterType)
                i++
            }
            newInstance = ctor.newInstance(*args)
        }
        return newInstance
    }

    private fun newEnumInstance(klazz: Class<*>): Any {
        return java.lang.Enum.valueOf(klazz as Class<out Enum<*>?>, "\$shadow_instance")!!
    }
}